molgenis-data-vcf/src/main/java/org/molgenis/data/vcf/utils/VcfUtils.java

```diff
@@ -83,337 +78,262 @@ public static String createId(Entity vcfEntity)
 			{
-				writer.write(value);
+				result.addAll(getAtomicAttributesFromList(attributeMetaData.getAttributeParts()));
 			}
 			else
 			{
-				writer.write('.');
+				result.add(attributeMetaData);
 			}
-			writer.write('\t');
 		}
+		return result;
+	}
 
-		boolean hasInfoFields = false;
-		// flexible 'info' field, one column with potentially many data items
-		for (AttributeMetaData attributeMetaData : vcfEntity.getEntityMetaData().getAttribute(INFO).getAttributeParts())
+	public static Map<String, AttributeMetaData> getAttributesMapFromList(Iterable<AttributeMetaData> outputAttrs)
+	{
+		Map<String, AttributeMetaData> attributeMap = new LinkedHashMap<>();
+		List<AttributeMetaData> attributes = getAtomicAttributesFromList(outputAttrs);
+		for (AttributeMetaData attributeMetaData : attributes)
 		{
-			String infoAttrName = attributeMetaData.getName();
-			if (attributesToInclude.isEmpty() || attributesToInclude.contains(infoAttrName))
-			{
-				if (attributeMetaData.getDataType().getEnumType() == FieldTypeEnum.BOOL)
-				{
-					Boolean infoAttrBoolValue = vcfEntity.getBoolean(infoAttrName);
-					if (infoAttrBoolValue != null && infoAttrBoolValue.booleanValue() == true)
-					{
-						writer.append(infoAttrName);
-						writer.append(';');
-						hasInfoFields = true;
-					}
-				}
-				else
-				{
-					String infoAttrStringValue = vcfEntity.getString(infoAttrName);
-					if (infoAttrStringValue != null)
-					{
-						writer.append(infoAttrName);
-						writer.append('=');
-						writer.append(infoAttrStringValue);
-						writer.append(';');
-						hasInfoFields = true;
-					}
-				}
-			}
+			attributeMap.put(attributeMetaData.getName(), attributeMetaData);
 		}
-		if (!hasInfoFields)
+		return attributeMap;
+	}
+
+	public static String toVcfDataType(MolgenisFieldTypes.FieldTypeEnum dataType)
+	{
+		switch (dataType)
 		{
-			writer.append('.');
+			case BOOL:
+				return VcfMetaInfo.Type.FLAG.toString();
+			case LONG:
+			case DECIMAL:
+				return VcfMetaInfo.Type.FLOAT.toString();
+			case INT:
+				return VcfMetaInfo.Type.INTEGER.toString();
+			case EMAIL:
+			case ENUM:
+			case HTML:
+			case HYPERLINK:
+			case STRING:
+			case TEXT:
+			case DATE:
+			case DATE_TIME:
+			case CATEGORICAL:
+			case XREF:
+			case CATEGORICAL_MREF:
+			case MREF:
+				return VcfMetaInfo.Type.STRING.toString();
+			case COMPOUND:
+			case FILE:
+			default:
+				throw new RuntimeException("unsupported vcf data type " + dataType);
 		}
+	}
 
-		// if we have SAMPLE data, add to output VCF
-		Iterable<Entity> sampleEntities = vcfEntity.getEntities(SAMPLES);
-		if (sampleEntities != null)
+	public static Iterator<Entity> reverseXrefMrefRelation(Iterator<Entity> annotatedRecords)
+	{
+		return new Iterator<Entity>()
 		{
-			boolean firstSample = true;
-			for (Iterator<Entity> it = sampleEntities.iterator(); it.hasNext();)
+			PeekingIterator<Entity> effects = Iterators.peekingIterator(annotatedRecords);
+
+			DefaultEntityMetaData resultEMD;
```

Comment:
It is already only created once. the "get" method returns the EMD if it exists, and creates it if it's not yet there